### PR TITLE
(Added)[JIRA-2751] - Adding post payment flow support for SEMOVI

### PR DESCRIPTION
**Pull request by rgrajalesmeli**
## What have changed?
Adding a new `PXPostPaymentConfiguration` in `MercadoPagoCheckoutBuilder` in order to support navigation to integrator post payment flow. 
This navigation to the post payment flow will be possible due to the implementation of a new series of notifications `MercadoPagoCheckout.NotificationCenter` that will tell the integrator whenever they should start their new custom flow and will provide a `resultBlock` where the integrator will be able to return a `PXBasePayment` to the PX flow in order to show a custom congrats.

New Clases Added in `MercadoPagoSDK`:
- `PXPostPaymentConfiguration`  Config class used in the builder to enable the post payment flow
- `MercadoPagoCheckout.NotificationCenter.SubscribeTo` This extension allows integrators to suscribe to post payment notification
- `MercadoPagoCheckout.NotificationCenter.PublishTo` This extension allows PX to post the post payment notifications
- `MercadoPagoCheckout.NotificationCenter.UnsubscribeTo` This extension allows integrators to unsubscribe observers for post payment flow

New Classes Added in `ExampleSwift` (Only used for testing):
- `CustomCheckoutWithPostPaymentViewController` New view controller that provides an entry point to the post payment flow
- `PostPaymentViewController` New view controller that will be called after getting the post payment notification
- `CustomPostPaymentProcessor` Custom processor used for adding post payment test cases 
 
## How to test:
The integration documentation can be found in the following link:
https://docs.google.com/document/d/1ND4qQlo_MmOfySHX-B5yB8PjIYhJhPxf1JSMPeVM5GU/edit#

## Prints for layout changes:
| SEMOVI Integrated in SP | SEMOVI Success | SEMOVI Failure | SEMOVI Payment Failure |
|:-:|:-:|:-:|:-:|
| SEMOVI_Integrated | SEMOVI_Test_iPhone_12 | SEMOVI_Test_iPhone_12_Failing_post_Payment | SEMOVI_Test_iPhone_12_Failing_Payment |

## Extras:
Current Unit tests are passing successfully but more will be added in a separate PR.
| Unit Tests  |
|:-:|
| Captura de Pantalla 2022-01-11 a la(s) 4 47 13 p  m |

### CODE REVIEW

File: ExampleSwift/ExampleSwift/Helpers/CustomPostPaymentProcessor.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        self.testCase = testCase
+    }
+
+    func startPayment(
+        checkoutStore: PXCheckoutStore,
+        errorHandler: PXPaymentProcessorErrorHandler,
+        successWithBusinessResult: @escaping ((PXBusinessResult) -> Void),
+        successWithPaymentResult: @escaping ((PXGenericPayment) -> Void)
+    ) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + 5.0, execute: { [self] in
```
Comment (matheusctMeli) [inline ExampleSwift/ExampleSwift/Helpers/CustomPostPaymentProcessor.swift:25]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/ExampleSwift/ExampleSwift/Helpers/CustomPostPaymentProcessor.swift#L25"></span>Trailing closure syntax should be used whenever possible.<br />
<code>trailing_closure</code> <code>CustomPostPaymentProcessor.swift:25</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/Config/PXPostPaymentConfiguration.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//  PXPostPaymentConfiguration.swift
+//  MercadoPagoSDKV4
+//
+//  Created by Ricardo Grajales Duque on 22/12/21.
+//
+
+import Foundation
+
+/**
+ Post payment configuration provides you support for being called in you own flow before showing the congrats screen and then go back to the PX flow.
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/Config/PXPostPaymentConfiguration.swift:11]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/Config/PXPostPaymentConfiguration.swift#L11"></span>Line should be 120 characters or less: currently 149 characters<br />
<code>line_length</code> <code>PXPostPaymentConfiguration.swift:11</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow+Services.swift [hunk 3]
```diff
@@ -102,7 +102,18 @@ extension PXPaymentFlow {
 
         model.shouldSearchPointsAndDiscounts = false
         let platform = MLBusinessAppDataService().getAppIdentifier().rawValue
-        model.mercadoPagoServices.getPointsAndDiscounts(url: PXServicesURLConfigs.MP_API_BASE_URL, uri: PXServicesURLConfigs.shared().MP_POINTS_URI, paymentIds: paymentIds, paymentMethodsIds: paymentMethodsIds, campaignId: campaignId, prefId: model.checkoutPreference?.id, platform: platform, ifpe: ifpe, merchantOrderId: model.checkoutPreference?.merchantOrderId, headers: headers, paymentTypeId: model.businessResult?.getPaymentMethodTypeId(), callback: { [weak self] pointsAndBenef in
+        model.mercadoPagoServices.getPointsAndDiscounts(url: PXServicesURLConfigs.MP_API_BASE_URL,
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow+Services.swift:105]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow+Services.swift#L105"></span>Multiline arguments should have their surrounding brackets in a new line.<br />
<code>multiline_arguments_brackets</code> <code>PXPaymentFlow+Services.swift:105</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift [hunk 4]
```diff
@@ -65,6 +67,26 @@ final class PXPaymentFlow: NSObject, PXFlow {
         }
     }
 
+    func goToPostPayment() {
+        PXNotificationManager.SuscribeTo.didFinishButtonAnimation(self, selector: #selector(showPostPayment))
+        PXNotificationManager.Post.animateButton(with: PXAnimatedButtonNotificationObject(status: "", interrupt: true))
+    }
+
+    @objc func showPostPayment() {
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift:75]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift#L75"></span>Attributes should be on their own lines in functions and types, but on the same line as variables and imports.<br />
<code>attributes</code> <code>PXPaymentFlow.swift:75</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift [hunk 5]
```diff
@@ -65,6 +67,26 @@ final class PXPaymentFlow: NSObject, PXFlow {
...
+        guard let notification = model.postPaymentNotificationName else {
+            return
+        }
+        MercadoPagoCheckout.NotificationCenter.PublishTo.postPaymentAction(withName: notification) { [unowned self] basePayment in
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift:79]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift#L79"></span>Line should be 120 characters or less: currently 130 characters<br />
<code>line_length</code> <code>PXPaymentFlow.swift:79</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlowModel.swift [hunk 6]
```diff
@@ -71,6 +75,14 @@ final class PXPaymentFlowModel: NSObject {
         return paymentResult == nil && businessResult == nil
     }
 
+    func needToGoToPostPayment() -> Bool {
+        let hasPostPaymentFlow = postPaymentNotificationName != nil
+        let paymentResultIsApproved = paymentResult?.isApproved() == true
+        let businessResultIsApprovedAndAccepted = businessResult?.isApproved() == true && businessResult?.isAccepted() == true
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlowModel.swift:81]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlowModel.swift#L81"></span>Line should be 120 characters or less: currently 126 characters<br />
<code>line_length</code> <code>PXPaymentFlowModel.swift:81</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift [hunk 7]
```diff
@@ -70,12 +87,13 @@ extension PXAnimatedButton: ProgressViewDelegate, CAAnimationDelegate {
...
-                                       height: animatedViewFrameInAnchorViewCoordinates.height)
-
-            let transitionAnimator = UIViewPropertyAnimator(duration: 0.5, dampingRatio: 1, animations: {
+            let toCircleFrame = CGRect(
+                x: animatedViewFrameInAnchorViewCoordinates.midX - animatedViewFrameInAnchorViewCoordinates.height / 2,
+                y: animatedViewFrameInAnchorViewCoordinates.minY,
+                width: animatedViewFrameInAnchorViewCoordinates.height,
+                height: animatedViewFrameInAnchorViewCoordinates.height
+            )
+            let transitionAnimator = UIViewPropertyAnimator(duration: Constants.defaultAnimationDuration, dampingRatio: 1, animations: {
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift:96]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift#L96"></span>Line should be 120 characters or less: currently 136 characters<br />
<code>line_length</code> <code>PXAnimatedButton.swift:96</code></td>
    </tr>
  </tbody>
</table>

Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift:96]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift#L96"></span>Trailing closure syntax should be used whenever possible.<br />
<code>trailing_closure</code> <code>PXAnimatedButton.swift:96</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift [hunk 8]
```diff
@@ -91,41 +109,65 @@ extension PXAnimatedButton: ProgressViewDelegate, CAAnimationDelegate {
...
+        iconImage.contentMode = .scaleAspectFit
+        iconImage.alpha = 0
+        animatedView.addSubview(iconImage)
+
+        UIView.animate(
+            withDuration: Constants.showIconAnimationDuration,
+            animations: {
                 iconImage.alpha = 1
-                iconImage.transform = CGAffineTransform(scaleX: 1.0, y: 1.0)
+                iconImage.transform = CGAffineTransform(scaleX: Constants.xScaleForIconAnimation, y: Constants.xScaleForIconAnimation)
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift:150]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift#L150"></span>Line should be 120 characters or less: currently 134 characters<br />
<code>line_length</code> <code>PXAnimatedButton.swift:150</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift [hunk 9]
```diff
@@ -91,41 +109,65 @@ extension PXAnimatedButton: ProgressViewDelegate, CAAnimationDelegate {
...
+                )
+            }
+        )
     }
 
+        private func expandAnimation(completion: @escaping () -> Void) {
+            guard let animatedView = self.animatedView else { return }
+
+            self.superview?.layer.masksToBounds = false
+            UIView.animate(withDuration: Constants.defaultAnimationDuration, animations: {
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift:166]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift#L166"></span>Arguments should be either on the same line, or one per line.<br />
<code>multiline_arguments</code> <code>PXAnimatedButton.swift:166</code></td>
    </tr>
  </tbody>
</table>

File: MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift [hunk 10]
```diff
@@ -91,41 +109,65 @@ extension PXAnimatedButton: ProgressViewDelegate, CAAnimationDelegate {
...
+                animatedView.transform = CGAffineTransform(scaleX: Constants.xScaleForExpandAnimation, y: Constants.xScaleForExpandAnimation)
```
Comment (matheusctMeli) [inline MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift:167]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/b35ebb5fadcd2007ddb7eafaacaa199e94c6bab2/MercadoPagoSDK/MercadoPagoSDK/UI/AnimatedButton/PXAnimatedButton.swift#L167"></span>Line should be 120 characters or less: currently 141 characters<br />
<code>line_length</code> <code>PXAnimatedButton.swift:167</code></td>
    </tr>
  </tbody>
</table>

**Comment by matheusctMeli** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">The PR title should follow the title convetion: (Added,Changed,Deprecated,Removed,Fixed,Security)[JIRA-XXX] - (brief description here)</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">SwiftLint also found 67 more violations with this PR.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">JIRA: <a href="https://mercadolibre.atlassian.net/browse/JIRA-2751">JIRA-2751</a></td>
    </tr>
  </tbody>
</table>

## Current coverage for MercadoPagoSDKV4.framework is `3.14%`
Files changed | - | - 
--- | --- | ---
PXPaymentFlowModel.swift | `0.00%` | :skull:
PXNotificationManager.swift | `0.00%` | :skull:
OneTapFlowModel.swift | `0.00%` | :skull:
MercadoPagoCheckout.swift | `0.00%` | :skull:
PXPaymentFlow.swift | `0.00%` | :skull:
MercadoPagoCheckoutViewModel.swift | `0.00%` | :skull:
MercadoPagoCheckout+NotificationManager.swift | `0.00%` | :skull:
LocationService.swift | `0.00%` | :skull:
PXAnimatedButton.swift | `0.00%` | :skull:
PXPaymentFlow+Services.swift | `0.00%` | :skull:
MercadoPagoCheckoutBuilder.swift | `20.00%` | :skull:

---

> Powered by [xcov](https://github.com/nakiostudio/xcov)

<p align="right" data-meta="generated_by_danger_pr">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: ExampleSwift/ExampleSwift/Controllers/CustomCheckoutWithPostPaymentViewController.swift [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+                paymentStatus: .REJECTED,
+                statusDetail: "cc_amount_rate_limit_exceeded"
+            )
+        case .error:
+            fatalError("genericPayment no debe ser invocado para este caso")
+        }
+    }
+}
+
+final class CustomCheckoutWithPostPaymentViewController: UIViewController {
```
Comment (matheusctMeli) [inline ExampleSwift/ExampleSwift/Controllers/CustomCheckoutWithPostPaymentViewController.swift:35]: <table data-meta="generated_by_danger_pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/mercadopago/px-ios/blob/7b3d78e2f168549a17a2bb5fd9451c17169d5f89/ExampleSwift/ExampleSwift/Controllers/CustomCheckoutWithPostPaymentViewController.swift#L35"></span>Type name should be between 3 and 40 characters long: 'CustomCheckoutWithPostPaymentViewController'<br />
<code>type_name</code> <code>CustomCheckoutWithPostPaymentViewController.swift:35</code></td>
    </tr>
  </tbody>
</table>

File: ExampleSwift/ExampleSwift/Resources/DebugConfig.xcconfig [hunk 12]
```diff
@@ -12,7 +12,9 @@
 // Use this file as example to generate needed config files
 
 //PX Environment - [alpha | beta | prod]
-PX_ENVIRONMENT=beta
+PX_ENVIRONMENT=prod
+//PX_ENVIRONMENT=beta
```
Comment (jscaramal-meli) [inline ExampleSwift/ExampleSwift/Resources/DebugConfig.xcconfig:16]: Keep xconfig changes local

File: MercadoPagoSDK/MercadoPagoSDK/Core/Checkout/MercadoPagoCheckout+NotificationManager.swift [hunk 13]
```diff
@@ -0,0 +1,53 @@
+//
+//  MercadoPagoCheckout+NotificationManager.swift
+//  MercadoPagoSDKV4
+//
+//  Created by Ricardo Grajales Duque on 22/12/21.
+//
```
Comment (jscaramal-meli) [inline MercadoPagoSDK/MercadoPagoSDK/Core/Checkout/MercadoPagoCheckout+NotificationManager.swift:6]: Remove file header

File: MercadoPagoSDK/MercadoPagoSDK/Config/PXPostPaymentConfiguration.swift [hunk 14]
```diff
@@ -0,0 +1,20 @@
+//
+//  PXPostPaymentConfiguration.swift
+//  MercadoPagoSDKV4
+//
+//  Created by Ricardo Grajales Duque on 22/12/21.
+//
```
Comment (jscaramal-meli) [inline MercadoPagoSDK/MercadoPagoSDK/Config/PXPostPaymentConfiguration.swift:6]: Remove file header

File: ExampleSwift/ExampleSwift/Controllers/PostPaymentViewController.swift [hunk 15]
```diff
@@ -0,0 +1,111 @@
+//
+//  PostPaymentViewController.swift
+//  ExampleSwift
+//
+//  Created by Ricardo Grajales Duque on 27/12/21.
+//  Copyright © 2021 Juan Sebastian Sanzone. All rights reserved.
+//
```
Comment (jscaramal-meli) [inline ExampleSwift/ExampleSwift/Controllers/PostPaymentViewController.swift:7]: Remove file header

File: MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift [hunk 16]
```diff
@@ -65,6 +67,52 @@ final class PXPaymentFlow: NSObject, PXFlow {
         }
     }
 
+    func goToPostPayment() {
+        PXNotificationManager.SuscribeTo.didFinishButtonAnimation(self, selector: #selector(showPostPayment))
+        PXNotificationManager.Post.animateButton(with: PXAnimatedButtonNotificationObject(status: "", interrupt: true))
+    }
+
+    @objc
+    func showPostPayment() {
```
Comment (ALEJOVILLA-ML) [inline MercadoPagoSDK/MercadoPagoSDK/Flows/PaymentFlow/PXPaymentFlow.swift:76]: should this be internal?

**Review by ALEJOVILLA-ML — APPROVED** [review]
LG

**Review by jlainog-meli — APPROVED** [review]

**Review by Esteban-ML — APPROVED** [review]

**Review by jscaramal-meli — APPROVED** [review]

**Pull request merged**